sdks/opik_optimizer/src/opik_optimizer/genetic_optimizer/genetic_optimizer.py

```diff
@@ -0,0 +1,1460 @@
+        Attempts chunk-level crossover first, then falls back to word-level.
+        """
+        str1_orig, str2_orig = str(ind1), str(ind2)
+
+        chunks1 = [chunk.strip() for chunk in str1_orig.split('.') if chunk.strip()]
+        chunks2 = [chunk.strip() for chunk in str2_orig.split('.') if chunk.strip()]
+
+        # Try chunk-level crossover if both parents have at least 2 chunks
+        if len(chunks1) >= 2 and len(chunks2) >= 2:
+            min_num_chunks = min(len(chunks1), len(chunks2))
+            # Crossover point is between 1 and min_num_chunks - 1
+            # This requires min_num_chunks >= 2, which is already checked.
+            point = random.randint(1, min_num_chunks - 1)
+            
+            child1_chunks = chunks1[:point] + chunks2[point:]
+            child2_chunks = chunks2[:point] + chunks1[point:]
+            
+            child1_str = '. '.join(child1_chunks) + ('.' if child1_chunks else '')
+            child2_str = '. '.join(child2_chunks) + ('.' if child2_chunks else '')
+            
+            return creator.Individual(child1_str), creator.Individual(child2_str)
+
+        # Fallback to word-level crossover if chunk-level is not suitable
+        words1 = str1_orig.split()
+        words2 = str2_orig.split()
+
+        # If either prompt is empty (no words), return parents
+        if not words1 or not words2:
+            return ind1, ind2
+
+        min_word_len = min(len(words1), len(words2))
+        # Need at least 2 words in the shorter prompt for a valid crossover point
+        if min_word_len < 2:
+            return ind1, ind2
+
+        # Crossover point for words: 1 to min_word_len - 1
+        point = random.randint(1, min_word_len - 1)
+        child1_words = words1[:point] + words2[point:]
+        child2_words = words2[:point] + words1[point:]
+        
+        return creator.Individual(' '.join(child1_words)), creator.Individual(' '.join(child2_words))
+
+    def _deap_mutation(
+            self,
+            individual: "creator.Individual",
+            task_config: TaskConfig
+        ) -> Tuple["creator.Individual",]:
+        """Enhanced mutation operation with multiple strategies. Requires task_config for some mutations."""
+        prompt = str(individual)
+        
+        # Choose mutation strategy based on current diversity
+        diversity = self._calculate_population_diversity()
+        if diversity < self.DEFAULT_DIVERSITY_THRESHOLD:
+            # Low diversity - use more aggressive mutations
+            mutation_type = random.random()
+
+            # Increased chance of semantic changes (includes radical)
+            if mutation_type < 0.5:
+                return self._semantic_mutation(prompt, task_config)
+            elif mutation_type < 0.8:
+                return self._structural_mutation(prompt)
+            else:
+                return self._word_level_mutation(prompt)
+        else:
```

Comment:
This if-branch is the same as the previous one, it's better to make the if-condition to determine the thresholds instead (0.4 vs 0.5, 0.7 vs 0.8) and not duplicate code.